src/CommonDCModelDropdown.php

```diff
@@ -212,14 +212,130 @@ public static function getSpecificValueToDisplay($field, $values, array $options
         return parent::getSpecificValueToDisplay($field, $values, $options);
     }
 
+    public function getItemRackForModel(): array
+    {
+        $itemtype = str_replace('Model', '', get_called_class());
+        return array_filter(
+            (new Item_Rack())->find([
+                'itemtype' => $itemtype,
+            ]),
+            // Filter out items that are not using the model
+            fn ($item_rack) => $itemtype::getById($item_rack['items_id'])->fields[strtolower(get_called_class()) . 's_id'] == $this->fields['id']
+        );
+    }
+
+    private function checkCellFilled(array $cell, int $orientation, int $hpos, float $depth): bool
+    {
+        if ($hpos == Rack::POS_NONE) {
+            return $this->checkCellFilled($cell, $orientation, Rack::POS_LEFT, $depth)
+                || $this->checkCellFilled($cell, $orientation, Rack::POS_RIGHT, $depth);
+        }
+
+        if (isset($cell[$hpos])) {
+            $accurateCell = array_slice($orientation ? array_reverse($cell[$hpos]) : $cell[$hpos], 0, $depth * 4);
+
+            if (in_array(1, $accurateCell)) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
     public function prepareInputForAdd($input)
     {
         return $this->managePictures($input);
     }
 
     public function prepareInputForUpdate($input)
     {
-        return $this->managePictures($input);
+        $input = $this->managePictures($input);
+
+        // Check if the model is used by an asset in a rack
+        // If so, check whether the new units required fit into the rack without modifying the positions
+        if (
+            isset($input['required_units']) && $input['required_units'] > $this->fields['required_units']
+            || isset($input['is_half_rack']) && $input['is_half_rack'] != $this->fields['is_half_rack']
+            || isset($input['depth']) && $input['depth'] != $this->fields['depth']
+        ) {
```

Comment:
The code use a bit too much nested loops and condition which we want to avoid if possible (https://en.wikibooks.org/wiki/Computer_Programming/Coding_Style/Minimize_nesting)

I would suggest:
* Isolate the code into a function (to keep `prepareInputForUpdate` light)
* Invert the first condition (if the input is OK -> return xxx; example here: https://www.jetbrains.com/help/resharper/InvertIf.html)
* Split the iterating on position + error reporting into two distinct parts (first part: build an array of position that need to be checked, second part: check each position and add error) -> this avoid duplicating the error logging code and reduce nesting a bit when handling the errors